### PR TITLE
fix: Icon position hidden when no icon selected #24068

### LINKED ISSUE #24068: [Task]: Icon position should be hidden when no icon is selected
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

image

The visibility of the position segment should be hidden when no icons are selected.

Created from https://github.com/appsmithorg/appsmith/pull/18533#issuecomment-1576580848

### PR DESCRIPTION
**Pull request by Harshithazemoso**
Fixes #24068  
_or_  
Fixes https://github.com/appsmithorg/appsmith/issues/24068
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags=""

### :mag: Cypress test results

> [!CAUTION]  
> If you modify the content in this section, you are likely to disrupt the CI result for your PR.

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

[Video:Fixing the bug](https://www.loom.com/share/02776c464df9467c9a6cf6d2e77b7945?sid=82025454-68ae-4acb-b3b2-916baab991fa)

### CODE REVIEW

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/Button/Button_spec.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ describe(
       cy.openPropertyPane("buttonwidget");
     });
 
-    it("1. Icon alignment should not change when changing the icon", () => {
+    it("1. Icon alignment should not change when changing the icon and the position setting should be hidden when icon field is none", () => {
```
[the remark this answers is not in the document]
Comment (Harshithazemoso) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/Button/Button_spec.js:25]: > Convert function expression to an arrow function for consistency and lexical scoping benefits.
> 
> ```diff
> -    it("1. Icon alignment should not change when changing the icon and the position setting should be hidden when icon field is none", function () {
> +    it("1. Icon alignment should not change when changing the icon and the position setting should be hidden when icon field is none", () => {
> ```
> 
> Committable suggestion

help @coderabbitai but its alreday arrow function right ,why are you asking to change that again

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/Others/MenuButton_spec.js [hunk 2]
```diff
@@ -11,7 +11,7 @@ describe(
       _.agHelper.AddDsl("menuButtonDsl");
     });
 
-    it("1. Icon alignment should not change when changing the icon", () => {
+    it("1. Icon alignment should not change when changing the icon and the position should be hidden when icon is none", () => {
```
[the remark this answers is not in the document]
Comment (Harshithazemoso) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/Others/MenuButton_spec.js:14]: > Convert function expression to an arrow function for consistency and lexical scoping benefits.
> 
> ```diff
> -    it("1. Icon alignment should not change when changing the icon and the position should be hidden when icon is none", function () {
> +    it("1. Icon alignment should not change when changing the icon and the position should be hidden when icon is none", () => {
> ```
> 
> Committable suggestion

@coerabbitai its already arrow function here also ,what is expected then ?

[the remark this answers is not in the document]
Comment (Harshithazemoso) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/Others/MenuButton_spec.js:14]: > > Seems like the humans are having a chat. I'll hop back into my burrow for now. If you need me again, just tag @coderabbitai in a new comment, and I'll come hopping out!

Help @coderabbitai ,in the code the function is already arrow function right ,so why are asking it to change ?

File: app/client/src/widgets/ButtonGroupWidget/widget/helpers.ts [hunk 3]
```diff
@@ -20,3 +20,21 @@ export const getStylesheetValue = (
...
+// for a button or menu item within a ButtonGroupWidget. It extracts the relevant button or menu item based on the provided property path and menu item flag.
+// The function returns true if the icon name is present and false if it is not  present.
+
+export const isIconNamePresent = (
+  props: ButtonGroupWidgetProps,
+  propertyPath: string,
+  isMenuItem = false,
+): boolean => {
+  const pathArray = propertyPath.split(".");
+  const buttonKey = pathArray[pathArray.length - (isMenuItem ? 4 : 2)];
```
Comment (jsartisan) [inline app/client/src/widgets/ButtonGroupWidget/widget/helpers.ts:34]: Hey, what are these values 4 and 2 signify? 🤔

Comment (jsartisan) [reply app/client/src/widgets/ButtonGroupWidget/widget/helpers.ts:34]: @Harshithazemoso

Comment (Harshithazemoso) [reply app/client/src/widgets/ButtonGroupWidget/widget/helpers.ts:34]: Hello @jsartisan The values 4 and 2 are used to correctly extract the part of the path that represents the buttonKey, depending on the structure of the propertyPath.
The 2 represents the index for non-menu item paths, focusing on simpler paths.
The 4 represents the index for menu item paths, which are assumed to be more nested.

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/Button/ButtonGroup_spec.js
Comment (jsartisan) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/Button/ButtonGroup_spec.js:1]: @Harshithazemoso  Hiding the iconAlign won't work if the iconNam is an binding. The reason is, that `PropertyControl` doesn't get the evaluated values. I tried it. So the solution you provided won't work for dynamic values.

Comment (Harshithazemoso) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/Button/ButtonGroup_spec.js:1]: Hello @jsartisan, but it's already implemented for other widgets binding to the `iconName` property, and it works fine there. I took reference from that implementation and wrote the code.
e.g for InputWidgetV2
Screenshot from 2024-06-03 15-59-50

Comment (jsartisan) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/Button/ButtonGroup_spec.js:1]: It does not work properly there as well. It's a limitation on our end right now 🥲 and we might not to able to prioritize changing our API for some time. So I am closing this PR for now as the solution does not work for all cases. Feel free to reopen the PR if you find an alternate solution.

Comment (Harshithazemoso) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/Button/ButtonGroup_spec.js:1]: > @Harshithazemoso Hiding the iconAlign won't work if the iconNam is an binding. The reason is, that `PropertyControl` doesn't get the evaluated values. I tried it. So the solution you provided won't work for dynamic values.

@jsartisan can please mention what you have meant by the dynamic values here.

Comment (riodeuno) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/Button/ButtonGroup_spec.js:1]: @Harshithazemoso Properties can be set dynamically using bindings. In this case, the Select Icon's property. 
In the case of dynamically provided values, the `hidden` function would require the "evaluated" value, as mentioned by @jsartisan. This is not enabled in the APIs today. The API only provides static user configured values.

**Pull request closed without merging**